src/server/index.js

```diff
@@ -1,188 +1,228 @@
-    rpc.writeProto(DialResponse, {})
-
-    pull(conn, shake.rest(), conn)
-  } catch (e) {
-    return rpc.writeProto(DialResponse, { error: Error.E_GENERIC })
-  }
-}
+const multiaddr = require('multiaddr')
+const PeerId = require('peer-id')
 
-const newMuxConn = (muxed) => prom(cb => muxed.newStream(cb))
-const sleep = (i) => new Promise((resolve, reject) => setTimeout(resolve, i))
+const sha5 = (data) => crypto.createHash('sha512').update(data).digest()
 
-const checkAckLoop = async (rpc, onEnd) => {
+const checkAckLoop = async (wrappedStream, onEnd) => {
   try {
     while (true) {
-      await rpc.readProto(DiscoveryAck)
-      await sleep(100)
+      await wrappedStream.readPB(DiscoveryAck)
+      await delay(100)
     }
   } catch (err) { // we'll get here after a disconnect
     onEnd()
   }
 }
 
+/**
+* Stardust Transport Server
+* @class
+*/
 class Server {
-  constructor ({ transports, addresses, muxers }) {
-    this.switch = new MicroSwitch({ transports, addresses, muxers, handler: this.handler.bind(this) })
-
+  /**
+   * @constructor
+   * @param {Object} options - Options for the listener
+   * @param {Array<multiaddr>} options.addresses
+   * @param {Array<Transport>} options.transports
+   * @param {Array<Multiplexer>} options.muxers
+   * @param {Array<Encryption>} options.encryption
+   * @param {PeerInfo} options.peerInfo
+   */
+  constructor ({
+    addresses = [multiaddr('/ip6/::/tcp/5892/ws')],
+    transports = [Transport],
+    muxers = [Muxer],
+    encryption = [Secio],
+    peerInfo
+  } = {}) {
+    this.peerAddr = addresses
     this.network = {}
     this.networkArray = []
-    this._emptyCachedDiscovery = this._cachedDiscovery = Buffer.from('01', 'hex')
+    this._cachedDiscovery = {}
 
-    this.switch.protos.addHandler('/p2p/stardust/0.1.0', (protocol, conn) => {
-      log('adding handler for stardust v0.1.0')
-
-      try {
-        this.handlerV010(conn, conn.info.info.muxed)
-      } catch (err) {
-        log(err)
-      }
-    })
+    this.libp2p = undefined
+    this._transports = transports
+    this._muxers = muxers
+    this._encryption = encryption
+    this._peerInfo = peerInfo
   }
 
-  async handler (conn) {
-    log('new connection')
+  removeFromNetwork (client) {
+    log('removing %s from network', client.id.toB58String())
+
+    delete this.network[client.id.toB58String()]
+    this.update()
+    this.broadcastDiscovery()
+  }
 
-    try {
-      const muxed = await this.switch.wrapInMuxer(conn, true) // add muxer ontop of raw socket
-      conn = await newMuxConn(muxed) // get a muxed connection
+  addToNetwork (connection, wrappedStream, id) {
+    log('adding %s to network', id.toB58String())
 
-      conn.muxed = muxed
-      await this.switch.negotiateProtocol(conn)
-    } catch (err) {
-      log(err)
+    const client = this.network[id.toB58String()] = {
+      id,
+      connection,
+      wrappedStream
     }
+
+    checkAckLoop(wrappedStream, () => this.removeFromNetwork(client))
```

Comment:
What is the value of the ack check? Right now after we broadcast the peers should respond with their ack messages. If we don't we're not disconnecting them, we just stop broadcasting to them. It seems like we should just pay more attention to the original stream. 

When a peer connects we perform a handshake. After the handshake we start streaming discovered peers to them. If the peer is not acting as a listener (they're just dialing through us), they can close the stream to not receive discovery events. Anytime a stream is closed we can remove that peer from our streams to pipe known peers to. This removes any need for an ack check. The ack check does add a "ping" mechansim to ensure peers are responding to us, but I don't know that that's necessary.

Reply:
TCP can leave dead connections sometimes for a really long time (especially if no data gets written by the still-alive side) if the other side vanishes (for example someone closing their laptop). So those closed-laptop zombie peers could be left around for minutes, unnecessarily polluting other peers discovery events (unless we're sending something to it, in which case the connection gets dropped quicker).

Also NATs tend to close connections that aren't sending data for some time

So if we dialed a peer through a relay with disabled discovery after we disconnect from that peer and dial another through the same realay after some time we'd end up with TCP errors because the NAT remove the connection pair from the table (so both sides still think the connection is open, but it's not)